Feature/static method route types

NigelGreenway opened the pull request:
Adds the ability for the following to be used:

``` php
Route::GET('/foo', 'foo');
Route::POST('/foo', 'foo');
Route::PUT('/foo', 'foo');
Route::PATCH('/foo', 'foo');
Route::DELETE('/foo', 'foo');
```

Allows the method call to be explicit where required, with the added validation of the method name.

If the method name does not exists an exception of `InvalidArgumentException` is thrown.

Commit:
Add initial HTTP methods array

Commit:
Add ability to call static method -

Allows `Route::[GET|POST|PUT|PATCH|DELETE|HEAD]` to be used for an
explicit method for the HTTP method name.

Commit:
Add missing  HTTP method name

AndrewCarterUK commented:
Thanks very much for this!

I definitely like the idea but I'm not too keen on the `_callStatic`. It's not helpful to IDEs (which can be solved with `@method` annotations but it seems like the main purpose of it here is to avoid the code duplication given the number of possible HTTP methods.

My inclination is that 99% of users are only interested in GET, POST, PUT and DELETE and that we can support other methods through the constructor anyway, so it might be better to implement those 4 just as static methods?

I'm probably persuadable on this, but I tend to stay away from a few of the magic methods in libraries (though I'm happy using them in apps). I find it harder to make the errors obvious and it can lead to unexpected results. In this example, if you only pass one parameter you get an 'undefined index' (from the list function) error rather than 'incorrect number of arguments' - which is more difficult to understand.

Interested to hear your thoughts - once we've settled on something I'll whip up some docs and merge :)

Thanks again!

NigelGreenway commented:
Hey Andrew,

I can move the magic method to allow for explicit static methods, shouldn't take long.

I will also have a look at the `list` issue too, bad on my part from not thinking about testing for that...

_Looks for shotgun..._ :smiley:

Commit:
Extract magic method into explicit methods -

As per feedback, I have created the explicit HTTP method static
methods.

refs: #1
authors: @NigelGreenway, @AndrewCarterUK

NigelGreenway commented:
I have left `PATCH` in there as well as I have used it previously and even though there are "debates" about it, I thought it might be worth leaving in, but have removed `HEAD` and `OPTIONS`. :)

Commit:
Change docbloc param type to

NigelGreenway commented:
I set it to `mixed` for the moment. Happy to change to `string|callback` if required.

AndrewCarterUK commented:
Thanks very much!
